platforms/core-execution/snapshots/src/main/java/org/gradle/internal/snapshot/impl/FileSystemSnapshotFilter.java

```diff
@@ -41,14 +41,18 @@ public class FileSystemSnapshotFilter {
     private FileSystemSnapshotFilter() {
     }
 
-    public static FileSystemSnapshot filterSnapshot(SnapshottingFilter.FileSystemSnapshotPredicate predicate, FileSystemSnapshot unfiltered) {
+    public static Optional<FileSystemLocationSnapshot> filterSnapshot(SnapshottingFilter filter, FileSystemLocationSnapshot unfiltered) {
+        if (filter.isEmpty()) {
+            return Optional.of(unfiltered);
+        }
+        SnapshottingFilter.FileSystemSnapshotPredicate predicate = filter.getAsSnapshotPredicate();
         DirectorySnapshotBuilder builder = MerkleDirectorySnapshotBuilder.noSortingRequired();
         AtomicBoolean hasBeenFiltered = new AtomicBoolean();
         unfiltered.accept(new RelativePathTracker(), new FilteringVisitor(predicate, builder, hasBeenFiltered));
         if (builder.getResult() == null) {
-            return FileSystemSnapshot.EMPTY;
+            return Optional.empty();
         }
-        return hasBeenFiltered.get() ? builder.getResult() : unfiltered;
+        return Optional.of(hasBeenFiltered.get() ? builder.getResult() : unfiltered);
     }
```

Comment:
💭 Instead of filtering `FileSystemSnapshot` (that can be empty, or contain snapshots from multiple roots) we now specifically filter the snapshot of a single location, and return exactly 0 or 1 snapshot matching the filter. Being more specific here makes it easier to work with the result.